mava/networks.py

```diff
@@ -0,0 +1,253 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+from typing import Dict, Sequence, Tuple, Union
+
+import chex
+import distrax
+import jax
+import jax.numpy as jnp
+import numpy as np
+from flax import linen as nn
+from flax.linen.initializers import constant, orthogonal
+
+from mava.types import (
+    Observation,
+    ObservationGlobalState,
+    RNNGlobalObservation,
+    RNNObservation,
+)
+
+
+class MLPTorso(nn.Module):
+    """MLP torso."""
+
+    layer_sizes: Sequence[int]
+    activation: str = "relu"
+    use_layer_norm: bool = False
+
+    def setup(self) -> None:
+        """Set up the activation function."""
+        if self.activation == "relu":
+            self.activation_fn = nn.relu
+        elif self.activation == "tanh":
+            self.activation_fn = nn.tanh
+
+    @nn.compact
+    def __call__(self, observation: chex.Array) -> chex.Array:
+        """Forward pass."""
+        x = observation
+        for layer_size in self.layer_sizes:
+            x = nn.Dense(
+                layer_size,
+                kernel_init=orthogonal(np.sqrt(2)),
+                bias_init=constant(0.0),
+            )(x)
+            if self.use_layer_norm:
+                x = nn.LayerNorm(use_scale=False)(x)
+            x = self.activation_fn(x)
+        return x
+
+
+class FeedForwardActor(nn.Module):
+    """Feedforward Actor Network."""
+
+    torso: nn.Module
+    num_actions: Sequence[int]
+
+    @nn.compact
+    def __call__(self, observation: Observation) -> distrax.Categorical:
+        """Forward pass."""
+        x = observation.agents_view
+
+        x = self.torso(x)
+        actor_output = nn.Dense(
+            self.num_actions, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(x)
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_output,
+            jnp.finfo(jnp.float32).min,
+        )
+
+        actor_policy = distrax.Categorical(logits=masked_logits)
+
+        return actor_policy
+
+
+class FeedForwardCritic(nn.Module):
+    """Feedforward Critic Network."""
+
+    torso: nn.Module
+    centralised_critic: bool = False
+
+    @nn.compact
+    def __call__(self, observation: Union[Observation, ObservationGlobalState]) -> chex.Array:
+        """Forward pass."""
+        if self.centralised_critic:
+            if not isinstance(observation, ObservationGlobalState):
+                raise ValueError("Global state must be provided to the centralised critic.")
+            # Get global state in the case of a centralised critic.
+            observation = observation.global_state
+        else:
+            # Get single agent view in the case of a decentralised critic.
+            observation = observation.agents_view
+
+        critic_output = self.torso(observation)
```

Comment:
`observation` -> `critic_input`:
```python
            # Get global state in the case of a centralised critic.
            critic_input = observation.global_state
        else:
            # Get single agent view in the case of a decentralised critic.
            critic_input = observation.agents_view
    critic_output = self.torso(critic_input)
```

Reply:
I think keeping it as observation because critic_input will be confused with critic_rnn_input